packages/react/src/ReactElement.js

```diff
@@ -158,6 +159,12 @@ const ReactElement = function(type, key, ref, self, source, owner, props) {
     _owner: owner,
   };
 
+  if (enableSpeculativeWork) {
+    // allows referential equality checking without having to hang onto old
+    // elements
+    element.residue = {};
```

Comment:
as mentioned elsewhere, will get rid of residue and replace with a childFiber like comparison implementation